File: Week 6/python/3.array_reverse_and_rotate.py
"""
WEEK 6 - PYTHON DSA
Topic: Array Reverse and Rotate
File: 3.array_reverse_and_rotate.py

CONCEPT:
    Two foundational in-place transformations on a list:
      1. Reverse a list using two pointers.
      2. Left/right rotate by k using the "reversal algorithm".

KEY POINTS:
    - Two-pointer reversal is O(n) time and O(1) extra space.
    - Reversal-based rotation is also O(n) time and O(1) space —
      far better than the naive O(n*k) shift loop.
    - Always reduce k modulo n; rotations of length n are no-ops.

ALGORITHM / APPROACH:
    Reverse subarray [l..r]:
        while l < r: swap, l++, r--
    Left rotate by k (k = k % n):
        reverse(arr, 0, k-1)
        reverse(arr, k, n-1)
        reverse(arr, 0, n-1)
    Right rotate by k = left rotate by (n - k % n).

PYTHON-SPECIFIC NOTES:
    - The "Pythonic" reverse is `arr.reverse()` or `arr[::-1]`.
    - Slice assignment can rotate in one expression:
        arr[:] = arr[k:] + arr[:k]   (uses O(n) extra space).
    - We still implement the manual algorithm because the topic is
      to learn the *technique*, not just the standard library.

DRY RUN:
    arr = [1,2,3,4,5]
        reverse: l=0,r=4 swap -> [5,2,3,4,1]
                 l=1,r=3 swap -> [5,4,3,2,1]
                 l=2,r=2 stop
    arr = [1,2,3,4,5], left rotate k=2
        rev(0,1): [2,1,3,4,5]
        rev(2,4): [2,1,5,4,3]
        rev(0,4): [3,4,5,1,2]

COMPLEXITY:
    Reverse : O(n) time, O(1) space
    Rotate  : O(n) time, O(1) space
"""

from typing import List


def reverse(arr: List[int], l: int, r: int) -> None:
    """In-place reverse of arr[l..r]."""
    while l < r:
        arr[l], arr[r] = arr[r], arr[l]
        l += 1
        r -= 1


def reverse_array(arr: List[int]) -> None:
    """In-place reverse of the whole array."""
    reverse(arr, 0, len(arr) - 1)


def left_rotate(arr: List[int], k: int) -> None:
    """In-place left rotation by k positions (reversal algorithm)."""
    n = len(arr)
    if n == 0:
        return
    k = k % n
    if k == 0:
        return
    reverse(arr, 0, k - 1)
    reverse(arr, k, n - 1)
    reverse(arr, 0, n - 1)


def right_rotate(arr: List[int], k: int) -> None:
    """Right rotate by k = left rotate by (n - k mod n)."""
    n = len(arr)
    if n == 0:
        return
    left_rotate(arr, n - (k % n))


def main() -> None:
    # --- Reverse ---
    arr1 = [1, 2, 3, 4, 5]
    print(f"Original: {arr1}")
    reverse_array(arr1)
    print(f"Reversed: {arr1}")

    # --- Left Rotate ---
    arr2 = [1, 2, 3, 4, 5]
    print(f"\nOriginal:         {arr2}")
    left_rotate(arr2, 2)
    print(f"Left Rotate by 2: {arr2}")

    # --- Right Rotate ---
    arr3 = [1, 2, 3, 4, 5]
    print(f"\nOriginal:          {arr3}")
    right_rotate(arr3, 2)
    print(f"Right Rotate by 2: {arr3}")

    # --- Edge cases ---
    single = [42]
    left_rotate(single, 5)
    print(f"\nSingle element rotated: {single}")

    empty: List[int] = []
    left_rotate(empty, 3)
    print(f"Empty array: {empty}")

    # Pythonic alternatives
    print("\nPythonic shortcuts:")
    print(f"reversed slice : {[1,2,3,4,5][::-1]}")
    arr = [1, 2, 3, 4, 5]
    k = 2
    print(f"slice rotation : {arr[k:] + arr[:k]}")


if __name__ == "__main__":
    main()


"""
NOTES — Python vs Java:
    - `a, b = b, a` performs a swap without a temporary variable.
    - Slicing (`arr[k:] + arr[:k]`) is concise but allocates O(n).
    - `list.reverse()` mutates in place; `reversed(list)` returns an iterator.
    - Java needs explicit length checks for empty arrays; Python's range/slice
      handle empty lists naturally.
"""

File: Week 6/java/3.ArrayReverseAndRotate.java
/*
 * WEEK 6 - ARRAYS
 * Topic: Reverse and Rotate an Array
 *
 * PROBLEM 1: Reverse an array in-place.
 * Input:  [1, 2, 3, 4, 5]
 * Output: [5, 4, 3, 2, 1]
 *
 * APPROACH: Two-pointer technique — swap arr[left] and arr[right], move inward.
 * Time: O(n)  Space: O(1)
 *
 * PROBLEM 2: Left-rotate an array by k positions.
 * Input:  [1, 2, 3, 4, 5], k=2
 * Output: [3, 4, 5, 1, 2]
 *
 * APPROACH (Reversal algorithm — O(n) time, O(1) space):
 * Step 1: Reverse first k elements:   [2,1, 3,4,5]
 * Step 2: Reverse remaining n-k:      [2,1, 5,4,3]
 * Step 3: Reverse entire array:       [3,4,5, 1,2]
 *
 * This is more efficient than the naive O(n*k) shift approach.
 */

import java.util.Arrays;

public class ArrayReverseAndRotate {

    // Reverse a subarray from index l to r (inclusive)
    static void reverse(int[] arr, int l, int r) {
        while (l < r) {
            int temp = arr[l];
            arr[l] = arr[r];
            arr[r] = temp;
            l++;
            r--;
        }
    }

    // Reverse entire array
    static void reverseArray(int[] arr) {
        reverse(arr, 0, arr.length - 1);
    }

    // Left rotate by k positions using reversal algorithm
    // Time: O(n), Space: O(1)
    static void leftRotate(int[] arr, int k) {
        int n = arr.length;
        k = k % n; // handle k >= n
        if (k == 0) return;
        reverse(arr, 0, k - 1);
        reverse(arr, k, n - 1);
        reverse(arr, 0, n - 1);
    }

    // Right rotate by k positions = left rotate by (n - k)
    static void rightRotate(int[] arr, int k) {
        int n = arr.length;
        leftRotate(arr, n - (k % n));
    }

    public static void main(String[] args) {
        // --- Reverse ---
        int[] arr1 = {1, 2, 3, 4, 5};
        System.out.println("Original: " + Arrays.toString(arr1));
        reverseArray(arr1);
        System.out.println("Reversed: " + Arrays.toString(arr1));

        // --- Left Rotate ---
        int[] arr2 = {1, 2, 3, 4, 5};
        System.out.println("\nOriginal:         " + Arrays.toString(arr2));
        leftRotate(arr2, 2);
        System.out.println("Left Rotate by 2: " + Arrays.toString(arr2));

        // --- Right Rotate ---
        int[] arr3 = {1, 2, 3, 4, 5};
        System.out.println("\nOriginal:          " + Arrays.toString(arr3));
        rightRotate(arr3, 2);
        System.out.println("Right Rotate by 2: " + Arrays.toString(arr3));

        // --- Edge cases ---
        int[] single = {42};
        leftRotate(single, 5);
        System.out.println("\nSingle element rotated: " + Arrays.toString(single));

        int[] empty = {};
        System.out.println("Empty array: " + Arrays.toString(empty));
    }
}
